Sources/NIOCore/VsockAddress.swift

```diff
@@ -0,0 +1,221 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2023 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+#if canImport(Darwin) || os(Linux)
+
+#if canImport(Darwin)
+import Darwin
+import CNIODarwin
+fileprivate let get_local_vsock_cid: @convention(c) (CInt) -> UInt32 = CNIODarwin_get_local_vsock_cid
+#elseif os(Linux)
+#if canImport(Glibc)
+import Glibc
+#elseif canImport(Musl)
+import Musl
+#endif
+import CNIOLinux
+fileprivate let get_local_vsock_cid: @convention(c) (CInt) -> UInt32 = CNIOLinux_get_local_vsock_cid
+#else
+#error("Unable to identify your C library.")
+#endif
+
+/// A vsock socket address.
+///
+/// A socket address is defined as a combination of a Context Identifier (CID) and a port number.
+/// The CID identifies the source or destination, which is either a virtual machine or the host.
+/// The port number differentiates between multiple services running on a single machine.
+///
+/// For well-known CID values and port numbers, see ``ContextID`` and ``Port`` respectively.
+public struct VsockAddress: Sendable {
+    private let _storage: Box<sockaddr_vm>
+
+    /// The libc socket address for a vsock socket.
+    public var address: sockaddr_vm { return _storage.value }
+
+    /// Get the context ID associated with the address.
+    public var cid: Int {
+        Int(Int32(bitPattern: self.address.svm_cid))
+    }
+
+    /// Get the port associated with the address.
+    public var port: Int {
```

Comment:
Why do these return `Int` instead of `ContextID` and `Port`?

Reply:
Relatedly, these two can be mutable.

Reply:
The user doesn't need to do that, just add an initializer that does it.

Reply:
Makes sense. However, do we even want to provide an initialiser from `Int`?

Both `ContextID` and `Port` have `init(rawValue: UInt32)` and `init(integerValue: UInt32)` to satisfy `RawRepresentable` and `ExpressibleByIntegerLiteral` respectively.

If we were to provide an init from `Int` how would we spell it to communicate the semantics that they'll be some conversion going on?

- `init?(_: Int)`
- `init(_: Int) throws`
- `init(truncatingIfNecessary: Int)`

Maybe the best thing to do is to _not_ provide one?

Reply:
There's no need to: we can simply crash, as the transformation from `Int` will do.

Reply:
Currently it feels like we have a safer API and, because we are conforming to `ExprissibleByIntegerLiteral` we have the ergonomics without the runtime crash:

```swift
_ = VsockAddress(cid: 4294967295, port: 12345)  // OK
_ = VsockAddress(cid: 4294967296, port: 12345)  // error: Integer literal '4294967296' overflows when stored into 'VsockAddress.ContextID'
```

Reply:
Oh sorry, I meant an init from `Int` on the `ContextID` and `Port`, not on anywhere else.

Reply:
I am indeed.